src/runner.rs

```diff
@@ -61,19 +68,107 @@ pub async fn run_with_config(
         .validate()?
         .build();
 
-    let (shutdown_tx, shutdown_rx) = watch::channel::<()>(());
+    let (shutdown_tx, shutdown_rx) = watch::channel(());
+    let (ready_tx, _ready_rx) = oneshot::channel();
     tokio::spawn(async move {
         // Don't unwrap in order to ensure that we execute
         // any subsequent shutdown tasks.
         signal::ctrl_c().await.ok();
         shutdown_tx.send(()).ok();
     });
 
-    if let Err(err) = server.run(shutdown_rx).await {
+    if let Err(err) = server.run(ready_tx, shutdown_rx).await {
         info!(log, "Shutting down with error"; "error" => %err);
         Err(Error::from(err))
     } else {
         info!(log, "Shutting down");
         Ok(())
     }
 }
+
+/// Start and run a proxy. Any passed in [`FilterFactory`]s are included
+/// alongside the default filter factories.
+pub async fn test(
+    base_log: slog::Logger,
+    mut testsuite: Testsuite,
+    filter_factories: impl IntoIterator<Item = DynFilterFactory>,
+) -> Result<(), Error> {
+    use once_cell::sync::Lazy;
+
+    static FEEDBACK_LOOP: Lazy<SocketAddr> = Lazy::new(|| {
+        let socket = std::net::UdpSocket::bind((Ipv4Addr::UNSPECIFIED, 0)).unwrap();
+        let local_addr = socket.local_addr().unwrap();
+        std::thread::spawn(move || loop {
+            let mut packet = [0; 0xffff];
+            let (_, addr) = socket.recv_from(&mut packet).unwrap();
+            let length = packet
+                .iter()
+                .position(|&x| x == 0)
+                .unwrap_or_else(|| packet.len());
+            let packet = &packet[..length];
+            eprintln!("Received: {:?}", packet);
```

Comment:
```suggestion
```
this seems leftover? or if we wanted to we can add an optional debug logging during tests